### PR TITLE
Improve support for `snapshot` on `default.mixed`

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Improves support for performing snapshots on `default.mixed`.

**Description of the Change:** Adds the ability to perform arbitrary state-based measurements. 

**Benefits:** Device will support `qml.Snapshot` fully.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by obliviateandsurrender** [comment]
[sc-60987]

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -286,6 +286,9 @@
 * The `qml.qchem.hf_state` function is upgraded to be compatible with the parity and Bravyi-Kitaev bases.
   [(#5472)](https://github.com/PennyLaneAI/pennylane/pull/5472)
 
+* `default.mixed` now supports arbitrary state-based measurements with `qml.Snapshots`.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:289]:
```suggestion
* `default.mixed` now supports arbitrary state-based measurements with `qml.Snapshot`.
```

File: pennylane/devices/default_mixed.py [hunk 2]
```diff
@@ -582,18 +593,80 @@ def _apply_density_matrix(self, state, device_wires):
...
+            snap_result = qml.math.mutual_info(
+                density_matrix,
+                indices0=wires0,
+                indices1=wires1,
+                c_dtype=self.C_DTYPE,
+                base=base,
+            )
+
+        else:  # pragma: no cover
+            raise DeviceError("Snapshots of {measurement} are not yet supported on default.mixed")
```
Comment (mudit2812) [inline pennylane/devices/default_mixed.py:647]:
```suggestion
            raise DeviceError(f"Snapshots of {type(measurement)} are not yet supported on default.mixed")
```

**Review by mudit2812 — COMMENTED** [review]
Looks good! Can you add a test to check that the correct error is raised with unsupported measurements? Otherwise ready to approve.

**Comment by astralcai** [comment]
@obliviateandsurrender Hi, we have a feature freeze for the v0.36 release starting next Monday. When you feel that this PR is ready to be merged, please label it merge-ready. I will be manually managing the merge queue over the next few days to make sure everything gets merged in time for feature freeze without wasting too much time on repeated CI runs.

**Pull request merged**